### ISSUE TITLE
"gradient" effect on DIVAnd maps[🐞]

### ISSUE THREAD

**Issue by qcazenave** [gher-uliege/DIVAnd.jl#129]
**Describe the bug**

![image](https://user-images.githubusercontent.com/118441818/232018429-4a695638-3a61-4c23-9b61-908548c175e8.png)
![image](https://user-images.githubusercontent.com/118441818/232018507-57f1b17e-5435-40ca-807a-2cde892e3306.png)
![image](https://user-images.githubusercontent.com/118441818/232018634-409f4614-8bad-4fa4-ac06-a4f067913b35.png)
![image](https://user-images.githubusercontent.com/118441818/232021243-626e719f-f01c-480f-9868-dcd60c21ae42.png)

** diva3d call **
`@time info = diva3d(
        (lonr,latr,depthr,TS),
        (obslon,obslat,obsdepth,obstime), obsval,
        len, epsilon2,                                           # epsilon2 = 1 (profile) or 10 (timeseries)
        outputfile[s], varname,
        bathname = bathfile,
        mask = mask,
        background = bkgd,                               #seasonal profile
        #plotres = plotres,
        transform = Anam.loglin(maxval),
        ncvarattrib = ncvarattrib,
        ncglobalattrib = metadata[s],
        surfextend = true,
        coeff_derivative2 = param_d2,                #param_2d = [0,0,0]
        memtofit = 50
);`

**Comment by jmbeckers** [comment]
I do not think it is a bug. Both situations arise in regions where you basically extrapolate and where a gradient in the nearest data can lead the extrapolation into unrealistic values.

In theory the error field should allow to identify such regions, in particular the "exact" error field generally is higher in such regions than the approximations used in practice (but is expensive).

Did you try the logit transformation instead of the loglin ? The latter might be responsible to allowing larger values than expected.

Just a stupid question: the values you use in DIVAnd for 2 µmole is 2 or 2E-6 ? (that could play a role when data are transformed)


Related note or question to developers: in loglin is is implicitely assumed that the data have values around 1 so that we have anomalies around zero? Maybe we need to add a scaling parameter too ?

**Comment by qcazenave** [comment]
Hi @jmbeckers ,
I agree, it is not really a bug but I was not sure about the label to use...
So, I haven't tried the logit transformation and the value I use for 2 umole is 2.

**Comment by qcazenave** [comment]
Update : logit instead of loglin = no change at all.
Use of logit : `transform = Anam.logit(min=minval,max=maxval)`

**Comment by jmbeckers** [comment]
that is strange. What are the values of minval and maxval ?

**Comment by qcazenave** [comment]
minval = 0
maxval = 800 for DIN and 400 for silicate (in umole/l)

**Comment by jmbeckers** [comment]
Isn't that very large ? What are the obs values in the deep region ?

**Comment by qcazenave** [comment]
Is it too large ? Should I keep the boundaries to the values found in the deep region ?

**Comment by ctroupin** [comment]
I would say that ideally one should not have to restrict the values of the input data, except if they are obviously wrong, which doesn't seem to be the case here.

What can be done is to use the residuals (diff. between observation and analysis at the location of the obs.) to discard some of the data points when the residual values is too large.

**Comment by qcazenave** [comment]
Yes, OK, but I don't understand how this can help with the effects presented above ?

**Comment by ctroupin** [comment]
Maybe by removing such data points you avoid the large gradients mentioned by @jmbeckers, so the extrapolation doesn't give so dramatic results.

**Comment by jmbeckers** [comment]
> I would say that ideally one should not have to restrict the values of the input data, except if they are obviously wrong, which doesn't seem to be the case here.
> 
> What can be done is to use the residuals (diff. between observation and analysis at the location of the obs.) to discard some of the data points when the residual values is too large.

 I did not mean to "cut" the **input** values. The logit transformation constrains the **analysis** to remain between two bounds.
So if you know that there are never values larger than 100, logit with an upper value of 100 will make the analysis always fall below 100 (and also will probably reduce some gradients and hence avoid some larger extrapolations).

**Comment by qcazenave** [comment]
Thank you for your answer @jmbeckers. Indead, it was not clear to me but I understand now.

**Comment by qcazenave** [comment]
Hi, I would like to get back to the **logit** transformation constraint on the values with the "min" and "max" parameters : for the **loglin** transformation, there is also a "max" parameter. I assumed it referred to the maximum value that would be authorized in the retrived field but the result of the analysis does not confirm my assumption. So my question is : what does this "max" parameter in the loglin transformation refer to ?

**Comment by jmbeckers** [comment]
Loglin:

Provide the following transform `log(x + epsilon)` (for x < t) and its inverse.
Beyond the threshold `t` (x ≥ t), the function is extended linearly in a
continuous way.

So for loglin beyond the max value a linear transformation is used. (and not a cutting/clipping, for that Logit is needed).

**Comment by qcazenave** [comment]
OK, my mistake, I forgot to re-check that.
Thank you for your answer.

**Comment by qcazenave** [comment]
With logit instead of loglin, I tried on chlorophyll-a data, using min=0mg/m2 and max=5mg/m3 and get the following error :
'''
ERROR: LoadError: DomainError with -2.0224948157370686:
log will only return a complex result if called with a complex argument. Try log(Complex(x)).
'''
On the same data, with loglin (using 5mg/m3 as threshold for the linear extension), no error.

**Comment by jmbeckers** [comment]
It probably means you have data outside of that range (there is no sanity check in the function before doing the anamorphis). Can you check on the input data that they are in the range ?

**Comment by qcazenave** [comment]
Yes, I most likely have data outside of that range. I thought it would work nevertheless taking into account the data according to the range.
